src/coreclr/src/vm/proftoeeinterfaceimpl.cpp

```diff
@@ -7030,7 +7032,184 @@ HRESULT ProfToEEInterfaceImpl::SetEnvironmentVariable(const WCHAR *szName, const
+                                                                       metadataBlob,
+                                                                       cbEventData,
+                                                                       eventData,
+                                                                       pActivityId,
+                                                                       pRelatedActivityId,
+                                                                       pEventThread,
+                                                                       numStackFrames,
+                                                                       stackFrames);
+            END_PIN_PROFILER();
+        };
+
+        EventPipeProviderConfiguration *pProviders = reinterpret_cast<EventPipeProviderConfiguration *>(pProviderConfigs);
+        UINT64 sessionID = EventPipe::Enable(NULL,
+                                             100, // circularBufferSizeInMB (shouldn't need this once it's synchronous)
+                                             pProviders,
+                                             cProviderConfigs,
+                                             EventPipeSessionType::Synchronous,
+                                             EventPipeSerializationFormat::NetTraceV4,
+                                             requestRundown,
+                                             NULL,
+                                             requestSamples,
+                                             eventPipeFunc);
+        EventPipe::StartStreaming(sessionID);
+
+        *pSession = sessionID;
+    }
+    EX_CATCH_HRESULT(hr);
+
+    return hr;
+#else // FEATURE_PERFTRACING
+    return E_NOTIMPL;
+#endif // FEATURE_PERFTRACING
+}
+
+HRESULT ProfToEEInterfaceImpl::EventPipeAddProviderToSession(
+        EVENTPIPE_SESSION session,
+        COR_PRF_EVENTPIPE_PROVIDER_CONFIG providerConfig)
+{
+
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_ANY;
+        EE_THREAD_NOT_REQUIRED;
+    }
+    CONTRACTL_END;
+
+    PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX(kP2EEAllowableAfterAttach | kP2EETriggers,
+        (LF_CORPROF,
+        LL_INFO1000,
+        "**PROF: EventPipeAddProviderToSession.\n"));
+
+#ifdef FEATURE_PERFTRACING
+    if (providerConfig.providerName == NULL)
+    {
+        return E_INVALIDARG;
+    }
+
+    HRESULT hr = S_OK;
+    EX_TRY
+    {
+        EventPipeSession *pSession = EventPipe::GetSession(session);
+        if (pSession == NULL)
+        {
+            hr = E_INVALIDARG;
+        }
+        else
+        {
+            EventPipeSessionProvider *pProvider = new EventPipeSessionProvider(
+                    providerConfig.providerName,
+                    providerConfig.keywords,
+                    (EventPipeEventLevel)providerConfig.loggingLevel,
+                    providerConfig.filterData);
+
+            pSession->AddSessionProvider(pProvider);
```

Comment:
I don't think there is any locking/protection around adding providers to a session while the session is active.  There might be dragons here 🐉  if you use this while a session is alive.

Reply:
Thanks, I copied this from the rundown logic but missed that it was holding the lock. I will make sure to grab the lock.